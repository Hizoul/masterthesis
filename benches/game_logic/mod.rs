pub mod tetromino;
pub mod field;
pub mod transposition;