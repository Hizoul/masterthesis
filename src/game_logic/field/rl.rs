use super::GameField;
use crate::heuristics::{heuristic_value_enemy_block,heuristic_touching,heuristic_value_connectability};
use rayon::prelude::*;
use rand::Rng;
use std::sync::RwLock;
use std::f64::MIN;
use rand::{thread_rng};

pub trait LearnHelper {
  fn update_reward(&mut self, player: i8, include_difference: bool) -> f64;
  fn basic_reward(&mut self, player: i8) -> f64;
  fn calculate_full_reward(&mut self, player: i8, include_difference: bool) -> f64;
  fn get_random_heuristic_play(&mut self) -> u8;
  fn get_best_heuristic_play(&mut self, true_randomness: bool) -> u8;
  fn get_best_heuristic_play_for_params(&mut self, connectability_weight: f64, enemy_block_wheight: f64, touching_weight: f64, score_weight: f64, true_randomness: bool, heuristic_type: u8) -> u8;
  fn get_other_player(&self, current_player: i8) -> i8;
}

impl LearnHelper for GameField {
  fn get_other_player(&self, current_player: i8) -> i8 {
    let next = current_player + 1;
    if next >= self.players {
      0
    } else {
      next
    }
  }
  fn update_reward(&mut self, player: i8, include_difference: bool) -> f64 {
    let mut new_reward = self.scores[player as usize] as f64; //(self.scores[player as usize] - self.scores[other as usize]) as f64;
    if include_difference {
      let other = self.get_other_player(player);
      new_reward -= self.scores[other as usize] as f64;
    }
    new_reward += heuristic_touching(self, player) * 0.3;
    new_reward /= 100.;
    new_reward
  }
  fn basic_reward(&mut self, player: i8) -> f64 {
    self.scores[player as usize] as f64
  }
  fn calculate_full_reward(&mut self, player: i8, include_difference: bool) -> f64 {
    let mut reward = 0.0;
    let indices = self.played_indices.clone();
    let mut shadow_field = self.empty_clone();
    for index in indices {
      let is_players_turn = shadow_field.current_player == player;
      shadow_field.place_block_using_play(index);
      if is_players_turn {
        reward += shadow_field.update_reward(player, include_difference);
      }
    }
    reward
  }
  fn get_random_heuristic_play(&mut self) -> u8 {
    let mut rng = thread_rng();
    self.get_best_heuristic_play_for_params(
      rng.gen_range(-1.0, 10.0),
      rng.gen_range(-1.0, 10.0),
      rng.gen_range(-1.0, 5.0),
      rng.gen_range(0.1, 25.0), true, 0)
  }
  fn get_best_heuristic_play(&mut self, true_randomness: bool) -> u8 {
    self.get_best_heuristic_play_for_params(0.3, 0.5, 1.0, 20.0, true_randomness, 0)
  }
  fn get_best_heuristic_play_for_params(&mut self, connectability_weight: f64, enemy_block_wheight: f64, touching_weight: f64, score_weight: f64, true_randomness: bool, heuristic_type: u8) -> u8 {
    let current_max_rwlock = RwLock::new(MIN);
    let player = self.current_player;
    let to_copy_from = &self;
    let best_index_rwlock: RwLock<Vec<u8>> = RwLock::from(Vec::with_capacity(175));
    (0..self.possible_plays.len()).into_par_iter().for_each(|current_index| {
      let mut field = to_copy_from.empty_clone();
      field.copy_from(to_copy_from);
      field.place_block_using_play(current_index as u8);
      let heuristic_value: f64 = match heuristic_type {
        1 => heuristic_touching(&field, player) + field.scores[player as usize] as f64,
        _ => (heuristic_value_connectability(&field, player) * connectability_weight)
      + (heuristic_value_enemy_block(&field, player) * enemy_block_wheight) + (heuristic_touching(&field, player) * touching_weight) +
      (field.scores[player as usize] as f64 * score_weight)
      };
      let mut current_max_writeable = current_max_rwlock.write().unwrap();
      let currmaxcmp = *current_max_writeable;
      if heuristic_value > currmaxcmp {
        let mut best_index_writeable = best_index_rwlock.write().unwrap();
        best_index_writeable.clear();
        best_index_writeable.push(current_index as u8);
        *current_max_writeable = heuristic_value;
      } else if heuristic_value == currmaxcmp {
        let mut best_index_writeable = best_index_rwlock.write().unwrap();
        best_index_writeable.push(current_index as u8);
      }
    });
    let mut best_index_writeable = best_index_rwlock.write().unwrap();
    best_index_writeable.sort();
    if best_index_writeable.len() == 0 {
      0 
    } else if best_index_writeable.len() == 1 {
      best_index_writeable[0]
    } else {
      best_index_writeable[if true_randomness {rand::thread_rng().gen_range(0, best_index_writeable.len()-1)} else {self.rng.gen_range(0, best_index_writeable.len()-1)}]
    }
  }
}