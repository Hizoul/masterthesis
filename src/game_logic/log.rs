use serde::{Deserialize, Serialize};

#[derive(Debug,Serialize,Deserialize,Clone,Copy)]
pub struct Play {
    pub x: i8,
    pub rotation: i8,
    pub block: i8,
    pub index: u8,
    pub global_index: u8
}

#[derive(Debug,Serialize,Deserialize,Copy,Clone)]
pub enum Payload {
  PayloadRolled {
    from: i8,
    block: i8
  },
  PayloadPlaced {
    x: i8,
    y: i8,
    block: i8,
    orientation: i8,
    from: i8
  },
  PayloadSkipped {
    from: i8,
    block: i8,
    reason: i8
  },
  PayloadConsidering {
    play_index: u8
  }
}

#[derive(Debug,Serialize,Deserialize,Clone)]
pub struct GameLog {
    pub log: Vec<Payload>
}

impl GameLog {
  pub fn new() -> GameLog {
    return GameLog {
      log: Vec::with_capacity(100)
    }
  }
  pub fn add_item(&mut self, item: Payload) {
    self.log.push(item)
  }
  pub fn block_placed(&mut self, user: i8, block_type: i8, block_orientation: i8, x: i8, y: i8) {
    self.add_item(Payload::PayloadPlaced {
      from: user,
      block: block_type,
      orientation: block_orientation,
      x: x,
      y: y
    })
  }
  pub fn rolled_block(&mut self, user: i8, block_type: i8) {
    self.add_item(Payload::PayloadRolled {
      from: user,
      block: block_type
    })
  }
  pub fn player_skipped(&mut self, user: i8, block_type: i8, reason: i8) {
    self.add_item(Payload::PayloadSkipped {
      from: user,
      block: block_type,
      reason: reason
    })
  }
  pub fn add_consideration(&mut self, play_index: u8) {
    self.add_item(Payload::PayloadConsidering{
      play_index
    })
  }
  pub fn get_current_roll(&self) -> (i8, i8) {
    for e in self.log.iter().rev() {
      match e {
        Payload::PayloadRolled {block,from} => {
          return (*from, *block)
        }
        _ => {}
      }
    }
    return (0, 0)
  }
  pub fn copy_from(&mut self, other: &GameLog) {
    self.log.clear();
    for entry in other.log.iter() {
      match entry {
        Payload::PayloadConsidering {play_index} => {
          self.add_item(Payload::PayloadConsidering{
            play_index: *play_index
          })
        }
        Payload::PayloadPlaced {block,from,x,y,orientation} => {
          self.block_placed(*from, *block, *orientation, *x, *y);
        }
        Payload::PayloadSkipped {block,from, reason} => {
          self.player_skipped(*from, *block, *reason);
        }
        Payload::PayloadRolled {block,from} => {
          self.rolled_block(*from, *block);
        }
      }
    }
  }
}