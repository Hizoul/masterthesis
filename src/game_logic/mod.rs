pub mod val;
pub mod cache;
pub mod tetromino;
pub mod log;
pub mod field;