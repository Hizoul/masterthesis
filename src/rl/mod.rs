pub mod gym_env;
pub mod policy_gradient;
pub mod aznet;
pub mod azmcts;
pub mod aztrain;