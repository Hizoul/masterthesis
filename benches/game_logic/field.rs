extern crate criterion;

use criterion::Criterion;
use rustyblocks::game_logic::field::{GameField};
use rustyblocks::game_logic::field::helper_structs::{GameRules};
use rustyblocks::game_logic::log::{GameLog};
use rustyblocks::game_logic::cache::{ShapeCache};
use std::sync::{Arc};
use rand::prelude::*;

const BLA: &'static str = "{\"log\":[{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":0,\"x\":8,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":0,\"x\":6,\"y\":0}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":0,\"x\":4,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":0,\"x\":7,\"y\":2}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":6,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":7,\"y\":3}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":6,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":6,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":6,\"y\":14}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":1,\"x\":7,\"y\":5}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":9,\"y\":3}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":3,\"x\":8,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":10,\"y\":6}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":2,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":8,\"y\":7}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":2,\"x\":4,\"y\":2}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":7,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":0,\"x\":4,\"y\":3}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":2,\"y\":6}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":3,\"x\":5,\"y\":5}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":3,\"x\":4,\"y\":6}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":3,\"x\":8,\"y\":11}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":1,\"x\":4,\"y\":9}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":3,\"x\":7,\"y\":12}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":1,\"x\":0,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":4,\"y\":12}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":9,\"y\":12}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":14}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":3,\"x\":0,\"y\":3}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":0,\"y\":4}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":1,\"x\":4,\"y\":14}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":7,\"x\":0,\"y\":5}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":7,\"x\":1,\"y\":8}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":16}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":4}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":3,\"x\":5,\"y\":15}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":8}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":9,\"y\":18}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":1,\"x\":2,\"y\":12}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":2,\"y\":15}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":5,\"x\":7,\"y\":15}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":2,\"x\":3,\"y\":16}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":3,\"x\":0,\"y\":10}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":17}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":4,\"y\":17}},{\"PayloadSkipped\":{\"from\":0,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":2,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":19}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":2,\"x\":7,\"y\":18}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":1,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadConsidering\":{\"play_index\":0}}]}";


struct Foo {
  counter: usize
}

impl Foo {
  pub fn add(&mut self) {self.counter+=1;}
  pub fn get(&self) -> usize {self.counter}
}

trait FooTrait {
  fn t_add(&mut self);
  fn t_get(&self) -> usize;
}

impl FooTrait for Foo {
  fn t_add(&mut self) {self.add();}
  fn t_get(&self) -> usize {self.get()}
}

pub fn bench_field(c: &mut Criterion) {
  let cache = ShapeCache::new();
  let rules = GameRules::default();
  let p = Arc::new(cache);
  let p2 = p.clone();
  let p3 = p.clone();
  let log: GameLog = serde_json::from_str(BLA).unwrap();
  c.bench_function("deterministic_play_until_over", move |b| b.iter(|| {
    let mut f = GameField::new_with_cache(2, p.clone(), rules);
    while !f.game_over {
      f.place_block_using_play(0);
    }
  }));
  c.bench_function("random_play_until_over", move |b| b.iter(|| {
    let mut  f = GameField::new_with_cache(2, p2.clone(), rules);
    while !f.game_over {
      let p = f.possible_plays.len() as u8;
      f.place_block_using_play(rand::thread_rng().gen_range(0,p));
    }
  }));
  c.bench_function("replay_log", move |b| b.iter(|| {
    let mut f = GameField::new_with_cache(2, p3.clone(), rules);
    f.restore_from_log(&log, false);
  }));
  c.bench_function("traitcost_direct", move |b| b.iter(|| {
    let mut f = Foo {counter:0};
    for _ in 0..99999999 {
      while f.get() < 9999999999 {
        f.add();
      }
    }
  }));
  c.bench_function("traitcost_struct", move |b| b.iter(|| {
    let mut f: Box<dyn FooTrait> = Box::new(Foo {counter:0});
    for _ in 0..99999999 {
      while f.t_get() < 9999999999 {
        f.t_add();
      }
    }
  }));
}