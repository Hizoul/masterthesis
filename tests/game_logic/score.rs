use rustyblocks::game_logic::field::{GameField};
use rustyblocks::game_logic::field::render::Renderable;
use rustyblocks::game_logic::field::rl::LearnHelper;
use rustyblocks::game_logic::log::{GameLog};
use insta::{assert_debug_snapshot,assert_snapshot};

const BLA: &'static str = "{\"log\":[{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":0,\"x\":1,\"y\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":4,\"y\":0}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":6,\"y\":0}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":2,\"y\":3}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":4,\"y\":2}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":5}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":4,\"x\":4,\"y\":4}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":9}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":4,\"y\":7}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":7,\"x\":0,\"y\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":3,\"x\":5,\"y\":7}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":0,\"x\":0,\"y\":5}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":0,\"x\":5,\"y\":8}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":3,\"y\":13}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":3,\"x\":4,\"y\":9}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":1,\"x\":1,\"y\":7}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":4,\"x\":7,\"y\":0}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":7,\"x\":0,\"y\":7}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":4,\"y\":11}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":8,\"y\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":7,\"y\":3}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":2,\"y\":8}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":8,\"y\":6}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":1,\"x\":0,\"y\":10}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":7,\"x\":7,\"y\":5}},{\"PayloadRolled\":{\"from\":1,\"block\":4}},{\"PayloadPlaced\":{\"from\":1,\"block\":4,\"orientation\":3,\"x\":0,\"y\":13}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":7,\"y\":8}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":3,\"x\":6,\"y\":9}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":2,\"x\":6,\"y\":12}},{\"PayloadRolled\":{\"from\":0,\"block\":4}},{\"PayloadPlaced\":{\"from\":0,\"block\":4,\"orientation\":4,\"x\":5,\"y\":13}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":4,\"y\":15}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":3,\"x\":7,\"y\":15}},{\"PayloadSkipped\":{\"from\":1,\"block\":4,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":2}},{\"PayloadPlaced\":{\"from\":1,\"block\":2,\"orientation\":3,\"x\":2,\"y\":14}},{\"PayloadRolled\":{\"from\":0,\"block\":3}},{\"PayloadPlaced\":{\"from\":0,\"block\":3,\"orientation\":1,\"x\":5,\"y\":18}},{\"PayloadRolled\":{\"from\":1,\"block\":0}},{\"PayloadPlaced\":{\"from\":1,\"block\":0,\"orientation\":1,\"x\":0,\"y\":14}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":0,\"x\":3,\"y\":17}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":1,\"y\":15}},{\"PayloadRolled\":{\"from\":1,\"block\":1}},{\"PayloadPlaced\":{\"from\":1,\"block\":1,\"orientation\":0,\"x\":6,\"y\":18}},{\"PayloadRolled\":{\"from\":0,\"block\":1}},{\"PayloadPlaced\":{\"from\":0,\"block\":1,\"orientation\":0,\"x\":8,\"y\":10}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadPlaced\":{\"from\":1,\"block\":3,\"orientation\":1,\"x\":8,\"y\":14}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadPlaced\":{\"from\":0,\"block\":2,\"orientation\":2,\"x\":2,\"y\":19}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadRolled\":{\"from\":0,\"block\":0}},{\"PayloadConsidering\":{\"play_index\":0}},{\"PayloadPlaced\":{\"from\":0,\"block\":0,\"orientation\":1,\"x\":8,\"y\":15}},{\"PayloadSkipped\":{\"from\":1,\"block\":0,\"reason\":1}},{\"PayloadSkipped\":{\"from\":0,\"block\":4,\"reason\":1}},{\"PayloadRolled\":{\"from\":1,\"block\":3}},{\"PayloadSkipped\":{\"from\":1,\"block\":3,\"reason\":0}},{\"PayloadRolled\":{\"from\":0,\"block\":2}},{\"PayloadSkipped\":{\"from\":0,\"block\":2,\"reason\":0}}]}";

#[test]
fn score_via_replay() {
  let log: GameLog = serde_json::from_str(BLA).unwrap();
  let mut field = GameField::new(2);
  field.restore_from_log(&log, true);
  assert_snapshot!("restored_field", &field.to_field_string());
  assert_debug_snapshot!("SCORES", field.scores);
  field.place_block_using_play(0);
  assert!(field.game_over == true);
  assert_snapshot!("full_reward", field.calculate_full_reward(0, false).to_string().as_str());
}