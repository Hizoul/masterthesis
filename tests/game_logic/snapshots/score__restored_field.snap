---
created: "2019-09-30T08:33:19.949236880Z"
creator: insta@0.8.1
source: tests/game_logic/score.rs
expression: "&field.to_field_string()"
---

0212004400
0222013420
4221222020
4204442220
4144342110
3434220230
3443212244
4404243404
4444124222
3444222212
4443121220
4434212120
3344112220
0444222210
3443222122
0044122222
4434222120
4044122044
3434222234
4340121122

